From: IrfanMohammad <irfan@graphiant.com>
Date: Sat, 2 Mar 2024 07:36:23 +0000
Subject: debug: serialize edit and data in sr_apply_changes

When sr_apply_changes fails, it is helpful to have the json for what
failed to apply. Without it reproducing the problem is time consuming.

If an crash happens, the json in memory is better to read than
going through lyd_node structs in gdb.

Closes: ENG-6310
---
 src/sysrepo.c | 21 +++++++++++++++++++++
 1 file changed, 21 insertions(+)

diff --git a/src/sysrepo.c b/src/sysrepo.c
index def05d4..dd9c1a4 100644
--- a/src/sysrepo.c
+++ b/src/sysrepo.c
@@ -4213,6 +4213,7 @@ sr_apply_oper_changes(struct sr_mod_info_s *mod_info, sr_session_ctx_t *session,
     const struct lyd_node *oper_edit;
     uint32_t mi_opts, i;
     int update_sm_data = 0;
+    char *edit_json = NULL, *data_json = NULL;
 
     assert(session && (session->ds == SR_DS_OPERATIONAL));
 
@@ -4233,6 +4234,10 @@ sr_apply_oper_changes(struct sr_mod_info_s *mod_info, sr_session_ctx_t *session,
         goto cleanup;
     }
 
+    /* remember the edit and data for debugging */
+    lyd_print_mem(&edit_json, oper_edit, LYD_JSON, LYD_PRINT_SHRINK | LYD_PRINT_WITHSIBLINGS);
+    lyd_print_mem(&data_json, mod_info->data, LYD_JSON, LYD_PRINT_SHRINK | LYD_PRINT_WITHSIBLINGS);
+
     /* generate the DS diff, data, and learn what modules are changed */
     if ((err_info = sr_modinfo_oper_ds_diff(mod_info, oper_edit))) {
         goto cleanup;
@@ -4285,6 +4290,11 @@ sr_apply_oper_changes(struct sr_mod_info_s *mod_info, sr_session_ctx_t *session,
     }
 
 cleanup:
+    if (err_info && data_json && edit_json) {
+        SR_LOG_WRN("Failed to apply oper changes: data %s edit %s", data_json, edit_json);
+    }
+    free(edit_json);
+    free(data_json);
     sr_release_data(old_oper_data);
     lyd_free_siblings(data_diff);
     lyd_free_siblings(old_oper_ds);
@@ -4298,6 +4308,7 @@ sr_apply_changes(sr_session_ctx_t *session, uint32_t timeout_ms)
     sr_error_info_t *err_info = NULL, *err_info2 = NULL;
     struct sr_mod_info_s mod_info;
     uint32_t mi_opts;
+    char *edit_json = NULL, *data_json = NULL;
 
     SR_CHECK_ARG_APIRET(!session || !SR_IS_STANDARD_DS(session->ds), session, err_info);
 
@@ -4333,6 +4344,9 @@ sr_apply_changes(sr_session_ctx_t *session, uint32_t timeout_ms)
         goto cleanup;
     }
 
+    lyd_print_mem(&edit_json, session->dt[session->ds].edit->tree, LYD_JSON, LYD_PRINT_SHRINK | LYD_PRINT_WITHSIBLINGS);
+    lyd_print_mem(&data_json, mod_info.data, LYD_JSON, LYD_PRINT_SHRINK | LYD_PRINT_WITHSIBLINGS);
+
     /* create diff */
     if ((err_info = sr_modinfo_edit_apply(&mod_info, session->dt[session->ds].edit->tree, 1, &err_info2))) {
         goto cleanup;
@@ -4354,12 +4368,19 @@ cleanup:
         /* free applied edit */
         sr_release_data(session->dt[session->ds].edit);
         session->dt[session->ds].edit = NULL;
+    } else {
+        if (data_json && edit_json) {
+            SR_LOG_WRN("Failed to apply changes data %s edit %s", data_json, edit_json);
+        }
     }
     if (err_info2) {
         /* return callback error if some was generated */
         sr_errinfo_merge(&err_info, err_info2);
     }
+    free(edit_json);
+    free(data_json);
     SR_LOG_DBG("Applying \"%s\" datastore changes %s.", sr_ds2str(session->ds), err_info ? "failed" : "success");
+
     return sr_api_ret(session, err_info);
 }
 
